dts-compatibility: dasharo-tools-suite: add Dasharo update via USB & iPXE

m-iwanicki opened the pull request.

m-iwanicki committed:
dts-compatibility: dasharo-tools-suite: add Dasharo update via USB & iPXE

Signed-off-by: Michał Iwanicki <michal.iwanicki@3mdeb.com>

m-iwanicki committed:
dts-compatibility: dasharo-tools-suite: disable BIOS lock

Signed-off-by: Michał Iwanicki <michal.iwanicki@3mdeb.com>

m-iwanicki commented:
@macpijan not sure how to test it. Currently the only ADL MSI connected doesn't work with external flashing (according to notes).
Can we somehow set `FLASHING_METHOD` variable just for this one platform?
MSI with RPL CPU doesn't seem to work with `v1.1.0` Dasharo (without DPP I can only update to `v1.1.1`).

Tested by commenting out `Flash Firmware` keyword

* [DTS009.001.tar.gz](https://github.com/user-attachments/files/17774609/DTS009.001.tar.gz)

    ```
    DEVICE_IP=192.168.10.72 RTE_IP=192.168.10.107 CONFIG=msi-pro-z690-a-wifi-ddr4 ./scripts/run.sh dasharo-compatibility/dasharo-tools-suite.robot -- -b cmd_logs.txt -t "DTS009.001*"
    ==============================================================================
    Dasharo-Tools-Suite
    ==============================================================================
    DTS009.001 Update Dasharo firmware by using DTS via USB works corr... | PASS |
    ------------------------------------------------------------------------------
    Since the asset has been checkout manually by you prior running this script, it will NOT be checked in automatically. Please return the device when work is finished.
    Dasharo-Tools-Suite                                                   | PASS |
    1 test, 1 passed, 0 failed
    ==============================================================================
    ```

* [DTS009.002.tar.gz](https://github.com/user-attachments/files/17774603/DTS009.002.tar.gz)

    ```
    DEVICE_IP=192.168.10.72 RTE_IP=192.168.10.107 CONFIG=msi-pro-z690-a-wifi-ddr4 ./scripts/run.sh dasharo-compatibility/dasharo-tools-suite.robot -- -b cmd_logs.txt -t "DTS009.002*"
    ==============================================================================
    Dasharo-Tools-Suite
    ==============================================================================
    DTS009.002 Update Dasharo firmware by using DTS via iPXE works cor... | PASS |
    ------------------------------------------------------------------------------
    Since the asset has been checkout manually by you prior running this script, it will NOT be checked in automatically. Please return the device when work is finished.
    Dasharo-Tools-Suite                                                   | PASS |
    1 test, 1 passed, 0 failed
    ==============================================================================
    ```

m-iwanicki committed:
dts-compatibility: dasharo-tools-suite: skip if FW_FILE is not defined

Signed-off-by: Michał Iwanicki <michal.iwanicki@3mdeb.com>

m-iwanicki committed:
dts-compatibility: dasharo-tools-suite: fix documentation

Signed-off-by: Michał Iwanicki <michal.iwanicki@3mdeb.com>

m-iwanicki committed:
dts-compatibility: dasharo-tools-suite: use dts-e2e keywords

Signed-off-by: Michał Iwanicki <michal.iwanicki@3mdeb.com>

m-iwanicki committed:
dts-lib.robot: use dts-boot instead of dts in shell

Signed-off-by: Michał Iwanicki <michal.iwanicki@3mdeb.com>

m-iwanicki committed:
dts-compatibility: dasharo-tools-suite: disable Intel ME

Signed-off-by: Michał Iwanicki <michal.iwanicki@3mdeb.com>

m-iwanicki committed:
dts-lib: add ME warning message variable

Signed-off-by: Michał Iwanicki <michal.iwanicki@3mdeb.com>

m-iwanicki committed:
lib/framework: enclose variables in triple quotes (""")

Signed-off-by: Michał Iwanicki <michal.iwanicki@3mdeb.com>

m-iwanicki committed:
dts-lib: Deal with Intel ME warning during update

Signed-off-by: Michał Iwanicki <michal.iwanicki@3mdeb.com>

m-iwanicki committed:
dasharo-compatibility: dts: Enter setup menu after update

To make sure that update went correctly

Signed-off-by: Michał Iwanicki <michal.iwanicki@3mdeb.com>

m-iwanicki committed:
dasharo-compatibility: dts: make DPP credentials optional

Signed-off-by: Michał Iwanicki <michal.iwanicki@3mdeb.com>

m-iwanicki committed:
dasharo-compatibility: dts: restore initial connection method after reboot

Signed-off-by: Michał Iwanicki <michal.iwanicki@3mdeb.com>

m-iwanicki committed:
lib/dts-lib: Read until `cpio.gz... ok`

Otherwise on PiKVM (e.g. MSI) it might take longer than 20 seconds to
boot into DTS which will result in trying to blindly enable SSH
before it's possible and failure when trying to connect via SSH

Signed-off-by: Michał Iwanicki <michal.iwanicki@3mdeb.com>

m-iwanicki commented:
Added more changes, now it's possible to pass DPP keys when running test, that way test will first load DPP credentials before trying to update. Resolves problem with being unable to update MSI as non-DPP versions are too old to tests on our platforms (and one that could work can't be flashed externally).
Also Update keyword can now deal with unflashable Intel ME region (doesn't happen with `1.1.3` binary but when trying to update from`1.1.2` (taken from cloud with enabled serial) DTS couldn't detect state of HAP disabled ME). Needs https://github.com/Dasharo/dts-scripts/pull/72 to be merged for it to work otherwise question will be asked twice and test will get stuck.

Some additional changes:
* `requirements.txt` - to fix errors while running tests or with pre-commit
* `Depends On` in `framework.robot` - using e.g.

    ```robot
    Depends On    ${TRUE}
    ...    Test can't run without setting FW_FILE, DPP_LOGS_KEY, DPP_DOWNLOAD_KEY and DPP_PASSWORD
    ```

    Results in:

    ```text
    Evaluating IF condition failed: Evaluating expression
    ''Test can't run without setting FW_FILE, DPP_LOGS_KEY, DPP_DOWNLOAD_KEY and DPP_PASSWORD' != None' failed:
    SyntaxError: unterminated string literal (detected at line 1) (<string>, line 1)
    ````
* `Boot Dasharo Tools Suite Via IPXE Shell` - if downloading and booting to DTS takes more than 20 seconds then `Boot Dasharo Tools Suite` will fail on PiKVM platforms when trying to blindly enable SSH before DTS booted. To fix this we should finish keyword only after last component is downloaded (which I'm assuming is `cpio.gz`)

---

Tested with local iPXE with DTS that included fix mentioned earlier. `msi_ms7d25_v1.1.3_ddr5.rom` was downloaded from cloud (official release) and modified to enable console redirection in BIOS via [dcuc](https://github.com/Dasharo/dcu):

```sh
./dcuc variable msi_ms7d25_v1.1.3_ddr5.rom  --set SerialRedirection --value Enabled
```

Run with

```sh
 robot -L TRACE -b cmd_logs.txt -v config:msi-pro-z690-a-ddr5 -v rte_ip:192.168.10.188 -v snipeit:yes \
    -v fw_file:msi_ms7d25_v1.1.3_ddr5.rom -v dpp_logs_key:'<logs_key>' -v dpp_download_key:'<download_key>' \
    -v dpp_password:'<dpp_password>' -v device_ip:192.168.10.39 -v boot_dts_from_ipxe_shell:True \
    -v dts_ipxe_link:http://<local_ip>:<local_port>/dts.ipxe -t "DTS009.002*" \
    dasharo-compatibility/dasharo-tools-suite.robot
```

```text
==============================================================================
Dasharo-Tools-Suite
==============================================================================
DTS009.002 Update Dasharo firmware by using DTS via iPXE works cor... | PASS |
------------------------------------------------------------------------------
Since the asset has been checkout manually by you prior running this script, it will NOT be checked in automatically. Please return the device when work is finished.
Dasharo-Tools-Suite                                                   | PASS |
1 test, 1 passed, 0 failed
==============================================================================
```

macpijan commented on dasharo-compatibility/dasharo-tools-suite.robot:
```diff
@@ -25,6 +25,8 @@ Test Setup          Run Keyword If    ${TESTS_IN_FIRMWARE_SUPPORT}
 
 
 *** Test Cases ***
+# After https://github.com/Dasharo/dts-scripts/pull/42 is released
```
It is still not released?
This MR is from 3 months ago

m-iwanicki replied:
6aa8557a1415ab9168921afec42ca80cac04eedb

m-iwanicki committed:
dasharo-tools-suite & dts-lib: Write without Enter in main menu

After Dasharo/dts-scripts#42 you can use options in main menu without
pressing Enter

Signed-off-by: Michał Iwanicki <michal.iwanicki@3mdeb.com>

PLangowski reviewed (approved):
lgtm, in the future we can think about checking the FW version after update (see if it changed/is higher

m-iwanicki committed:
dasharo-tools-suite: update variable check to new format

Signed-off-by: Michał Iwanicki <michal.iwanicki@3mdeb.com>

m-iwanicki commented:
Rebased and tested to make sure my changes didn't affect e2e tests: https://github.com/Dasharo/meta-dts/actions/runs/14242520046/job/39916969502
